packages/graph/hash_graph/lib/graph/src/api/rest/middleware.rs

```diff
@@ -74,3 +83,120 @@ where
 
     Ok(bytes)
 }
+
+struct HeaderExtractor<'a>(&'a http::HeaderMap);
+// Let OpenTelemetry pick the field names to make our headers "standardized".
+// We would have to set `traceparent` in a header to correlate spans.
+impl<'a> Extractor for HeaderExtractor<'a> {
+    fn get(&self, key: &str) -> Option<&str> {
+        self.0.get(key).and_then(|value| value.to_str().ok())
+    }
+
+    fn keys(&self) -> Vec<&str> {
+        self.0
+            .keys()
+            .map(hyper::header::HeaderName::as_str)
+            .collect()
+    }
+}
+
+fn extract_header_remote_context(headers: &http::HeaderMap) -> opentelemetry::Context {
+    let extractor = HeaderExtractor(headers);
+    let ctx =
+        opentelemetry::global::get_text_map_propagator(|propagator| propagator.extract(&extractor));
+
+    if ctx.span().span_context().is_valid() {
+        // Remote context where the trace is correlated.
+        ctx
+    } else {
+        // New, local context with generated ids.
+        ctx.with_remote_span_context(SpanContext::new(
+            RandomIdGenerator::default().new_trace_id(),
+            SpanId::INVALID,
+            ctx.span().span_context().trace_flags(),
+            // explicitly make it non-remote
+            false,
+            ctx.span().span_context().trace_state().clone(),
+        ))
+    }
+}
+
+fn parse_x_forwarded_for(headers: &http::HeaderMap) -> Option<Cow<'_, str>> {
+    let value = headers.get("x-forwarded-for")?;
+    let value = value.to_str().ok()?;
+    let mut ips = value.split(',');
+    Some(ips.next()?.trim().into())
+}
+
+// Based on https://github.com/tokio-rs/axum/pull/769
+pub fn span_maker(request: &hyper::Request<hyper::Body>) -> tracing::Span {
```

Comment:
We can also specify `on_failure` and `on_response` to populate error codes and server faults